contracts/zaps/steth/ZapSteth.sol

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity 0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "../../../interfaces/curve/Curve.sol";
+import "../../../interfaces/lido/ISteth.sol";
+
+import "@openzeppelin/contracts/math/Math.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import {
+    SafeERC20,
+    SafeMath,
+    IERC20,
+    Address
+} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "../BaseZap.sol";
+
+contract ZapSteth is BaseZap {
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using SafeMath for uint256;
+
+    ISteth public stETH =
+        ISteth(address(0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84));
+
+    //slippage allowance is out of 10000. 20 is 2%
+    ICurveFi public StableSwapSTETH =
+        ICurveFi(address(0xDC24316b9AE028F1497c275EB9192a3Ea0f67022));
+
+    uint256 public constant DEFAULT_SLIPPAGE = 50;
+    bool private _noReentry = false;
+
+    constructor() public Ownable() {
+        want = IERC20(address(0x06325440D014e39736583c165C2963BA99fAf14E));
+        yVault = IYVault(address(0xdCD90C7f6324cfa40d7169ef80b12031770B4325));
+
+        want.safeApprove(address(yVault), uint256(-1));
+        want.safeApprove(address(StableSwapSTETH), uint256(-1));
+        stETH.approve(address(StableSwapSTETH), uint256(-1));
+    }
+
+    //we get eth
+    receive() external payable {
+        if (_noReentry) {
+            return;
+        }
+        _zapEthIn(DEFAULT_SLIPPAGE);
+    }
+
+    function zapEthIn() external payable override {
+        if (_noReentry) {
+            return;
+        }
+        _zapEthIn(DEFAULT_SLIPPAGE);
+    }
+
+    function zapEthIn(uint256 slippageAllowance) external payable override {
+        if (_noReentry) {
+            return;
+        }
+
+        _zapEthIn(slippageAllowance);
+    }
+
+    function _zapEthIn(uint256 slippageAllowance) internal {
+        uint256 balanceBegin = address(this).balance;
+
+        if (balanceBegin < 2) return;
+
+        uint256 halfBal = balanceBegin.div(2);
+
+        //test if we should buy instead of mint
+        uint256 out = StableSwapSTETH.get_dy(0, 1, halfBal);
+
+        if (out < halfBal) {
+            stETH.submit{value: halfBal}(owner());
+        }
+
+        uint256 balanceMid = address(this).balance;
+        uint256 balance2 = stETH.balanceOf(address(this));
+
+        StableSwapSTETH.add_liquidity{value: balanceMid}(
+            [balanceMid, balance2],
+            0
+        );
+
+        uint256 outAmount = want.balanceOf(address(this));
+
+        require(
+            outAmount.mul(slippageAllowance.add(10000)).div(10000) >=
+                balanceBegin,
+            "TOO MUCH SLIPPAGE"
+        );
+
+        yVault.deposit(outAmount, msg.sender);
+    }
+
+    function zapStEthIn(uint256 stEthAmount, uint256 slippageAllowance)
```

Comment:
this is the function that can be generalized with `zapin` that accepts an erc20 token, otherwise the UI will have a lot of issues keeping up with the interfaces across all zaps, every zap will have a diff name per supported erc20 which doesnt scale

Reply:
I tough we would have the EthZAP and the ERC20 zap.

Reply:
i think its less gas and less work for the UI to manage a zap contract per vault that supports ERC20s and ETH

Reply:
From the conversation with @fameal we are starting on ETH zaps for now and not ERC20.